[Erlang] initial port of basics exercise from C#

bgottlob opened the pull request.

bgottlob committed:
[Erlang] initial port of basics exercise from C#

bgottlob committed:
[Erlang] Fix config.json formatting

bgottlob committed:
[Erlang] Run prettier over *.json and *.md files

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/after.md:
```diff
@@ -0,0 +1,8 @@
+See the following sections of the Erlang documentation for more information:
```
The after.md document should contain all information that is introduced in the introduction.md document (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd)).

bgottlob replied:
Should I paste in the information from `introduction.md` into `after.md`, or should I attempt to summarize it here?

ErikSchierboom replied:
We've since updated the spec and the `after.md` is now replaced by a concept `about.md` document. For more information, see https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#concept-files

In short: the `after.md` document is no longer part of the exercise. Instead, each exercise will have a `concepts/<slug>/about.md` and `concepts/<slug>/links.json` file.

> Should I paste in the information from introduction.md into after.md, or should I attempt to summarize it here?

The docs can probably best explain this: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#aboutmd

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/hints.md:
```diff
@@ -0,0 +1,3 @@
+- You can use [arithmetic operators](http://erlang.org/doc/reference_manual/expressions.html#arithmetic-expressions) to implement the functions
```
These hints are nice!

ErikSchierboom replied:
When possible, links are linked to individual tasks. If that doesn't make sense for a particular hint, they can be a general hint. The format in which the hints.md file should be written is slightly different, where there are headings added (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docshintsmd)). For an example of such a hints file, see [this file](https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/floating-point-numbers/.docs/hints.md).

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,60 @@
+calculator:add(1, 3).
+```
+
+Single line comments are preceded by `%`. Erlang does not support multi-line comments. A common practice is to start single-line comments with `%%` and to precede comments on the same line as code with `%`.
````
In trying to keep the introduction.md as minimal as possible, maybe remove the "A common practice is to start single-line comments with `%%` and to precede comments on the same line as code with `%`." bit and only have that in the `after.md` document?

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
```diff
@@ -0,0 +1,60 @@
+
+Variable names must begin with an uppercase letter and are generally written in camel case.
+
+To enforce [immutabilty][immutability], once a variable is bound to a value, it cannot be updated:
```
There is a link to a resource here but I don't think that link is declared anywhere. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd):

> Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

In this case, I think it is fine to mention immutability (maybe using underscores to make it italic), but not to link to an external resources. Linking to external resources is great for `after.md` documents though, so please do add it there.

ErikSchierboom replied:
The above comment on not preferrably not using links in the introduction.md document also applies to the other links in this file.

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,60 @@
+% exception error: no match of right hand side value 6
+```
+
+The `=` operator is also used for performing [pattern matching][pattern_matching] to bind values to variables. In the case above, `Count` is bound to the value `5`, then pattern matching is performed with `6` on the right side being matched to `5` on the left. `5` is not matched with `6`, so an error occurs. [Pattern matching][pattern_matching] will be discussd in more detail in later exercises.
````
```suggestion
The `=` operator is also used for performing [pattern matching][pattern_matching] to bind values to variables. In the case above, `Count` is bound to the value `5`, then pattern matching is performed with `6` on the right side being matched to `5` on the left. `5` is not matched with `6`, so an error occurs. [Pattern matching][pattern_matching] will be discussed in more detail in later exercises.
```

ErikSchierboom replied:
In trying to keep the introduction.md as minimal as possible, I'm wondering if this section could maybe removed here and have it only in the after.md document? Note that the introduction document's goal is to provide information that gives the student just enough context to figure out the solution themselves. Anything that is not essential for the student to be able to solve the solution should be omitted from the introduction.md (it can be included in the after.md document though).

NobbZ replied:
I'm also in favor of droping this paragraph and leaving the error as is, just showing that rebinding isn't possible.

For pattern matching there should be a distinct concept exercise that then explains what happens in the example above.

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,60 @@
+add(1, 3).
+```
+
+To allow for a function to be invoked outside of its module, use `export` directive to specify a list of exported functions containing each function's name and arity:
````
```suggestion
To allow for a function to be invoked outside of its module, use the `export` directive to specify a list of exported functions containing each function's name and arity:
```

NobbZ replied:
```suggestion
To allow for a function to be invoked outside of its module, use the `export` module attribute to specify a list of exported functions containing each function's name and arity:
```

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,60 @@
+-export([add/2, double/1]).
+```
+
+Note that two functions with the same name but different arities are two _different_ functions. Erlang does not have a concept of overloaded functions.
````
In trying to keep the introduction.md as minimal as possible, maybe remove this line here and have it only in the after.md document?

ErikSchierboom commented on languages/erlang/exercises/concept/basics/.meta/design.md:
```diff
@@ -0,0 +1,29 @@
+
+The concepts this exercise unlocks are:
+
+TBD
```
This will likely be:

```suggestion
- `basics`
```

This is how most tracks organize this, as subsequent exercises can then simply have `basics` as a prerequisite.

ErikSchierboom reviewed (approved):
Awesome work! I've left some comments. 

The PR should also add an entry to the `config.json` file: https://github.com/exercism/v3/blob/master/languages/erlang/config.json#L12

NobbZ commented on languages/erlang/exercises/concept/basics/.docs/introduction.md:
```diff
@@ -0,0 +1,60 @@
+
+The `=` operator is also used for performing [pattern matching][pattern_matching] to bind values to variables. In the case above, `Count` is bound to the value `5`, then pattern matching is performed with `6` on the right side being matched to `5` on the left. `5` is not matched with `6`, so an error occurs. [Pattern matching][pattern_matching] will be discussd in more detail in later exercises.
+
+Erlang is a [functional language][functional] and requires all functions to be defined in a _module_. The `module` directive is used to define a module.
```
```suggestion
Erlang is a [functional language][functional] and requires all functions to be defined in a _module_. The `module` module attribute is used to define a modules name.
```

NobbZ commented on languages/erlang/exercises/concept/basics/.docs/hints.md:
```diff
@@ -0,0 +1,3 @@
+- You can use [arithmetic operators](http://erlang.org/doc/reference_manual/expressions.html#arithmetic-expressions) to implement the functions
+- Remember that the last expression _evaluated_ in the function body will be returned
+- Remember that [variables](http://erlang.org/doc/reference_manual/expressions.html#variables) must start with an uppercase letter
```
*YES*! This is a good one, and fools me every now and then, when I do some Erlang after a streak of Python, Elixir or Go…

NobbZ commented:
I've put some comments.

Especially I think the "directive" vs "module attribute" should be fixed, as the Erlang manual refers to both as examples of "modules attributes":

http://erlang.org/doc/reference_manual/modules.html

> ```erl
> -module(m).          % module attribute
> -export([fact/1]).   % module attribute
>
> fact(N) when N>0 ->  % beginning of function declaration
>     N * fact(N-1);   %  |
> fact(0) ->           %  |
>     1.               % end of function declaration
> ```

ErikSchierboom commented:
Thanks @NobbZ! It is indeed important to use the correct naming for terms which should match official documentation and such.

bgottlob committed:
[Erlang] Fix basics exercise after review

- Move additional infromation from introduction.md to after.md
- Adopt "Pascal case" and "module attribute" terminology
- Remove unnecessary links from introduction.md
- Add entry to Erlang's config.json

bgottlob commented:
Thanks for the review @ErikSchierboom and @NobbZ ! I agreed with all of your suggestions and made changes accordingly in https://github.com/exercism/v3/pull/1523/commits/2edbcbb5fc59f682b176ee5b445e7b60e2e77c6b. Once I get some clarity on [this comment](https://github.com/exercism/v3/pull/1523#discussion_r523891912), I will finish this up.

ErikSchierboom commented on languages/erlang/config.json:
```diff
@@ -5,7 +5,15 @@
-    "concept": [],
+    "concept": [
+      {
+        "slug": "basics",
+        "uuid": "e21279ba-ee09-412a-902d-478ccdc5d28c",
+        "prerequisites": [],
+        "topics": ["math", "functions", "variables"],
+        "difficulty": 1
```
We've updated the spec a bit in the meantime. Here's how to update to the latest specs

```suggestion
        "slug": "lucians-luscious-lasagna",
        "name": "Lucian's Luscious Lasagna",
        "uuid": "e21279ba-ee09-412a-902d-478ccdc5d28c",
        "concepts": ["basics"],
        "prerequisites": [],
```

As we're now naming exercises for their story/theme (`lucians-luscious-lasagna` in this case), the directory of the exercise should also be changed to `lucians-luscious-lasagna` (so that would be `languages/erlang/exercises/concept/lucians-luscious-lasagna`).

ErikSchierboom commented on languages/erlang/config.json:
```diff
@@ -5,7 +5,15 @@
+        "difficulty": 1
+      }
+    ],
     "practice": []
```
With the changing spec, we've also introduced a new top-level object: `concepts`. All the concepts listed in the exercise's `concepts` key should have a corresponding entry in the `concepts` key. Taking into account the suggestion added in the above comment, this would mean adding:

```suggestion
    "practice": []
    "concepts": [
      {
        "uuid": "2eb4a463-355f-46ef-ac55-a75ec5afdf86",
        "slug": "basics",
        "name": "Basics",
        "blurb": "TODO: add blurb for basics concept"
    }
  ]
```

The `blurb` can be updated later,

ErikSchierboom reviewed (approved):
Great! I've left some comments.

bgottlob committed:
[Erlang] Update basics exercise for new specs

- Add basics to "concepts" list in config.json
- Update config.json with theme based slug
- Move .docs/after.md to about.md and include introduction info
- Move external links to links.json

bgottlob committed:
[Erlang] Fix config.json merge conflict

bgottlob commented:
I fixed the merge conflict in `config.json` and made your suggested changes to meet the new specs.

ErikSchierboom commented:
This Friday (Jan 29th 2021), we're going to explode this repo back into the individual track repositories and archive this repo.

As such, this PR needs to be merged by Friday 9am UTC, or it will need to be manually reopened against the track repo.

NobbZ commented:
From the erlang point of view, I think this can be merged as is, though it seems as if there are still some validations failing on the configuration side of things.

As this would be the only converted exercise anyway, I have no priority to get this merged, but instead might cherry pick parts of this PR later on when I have more time to focus on gettting the erlang track to v3… Perhaps after the pandemic…

ErikSchierboom commented:
> As this would be the only converted exercise anyway, I have no priority to get this merged, but instead might cherry pick parts of this PR later on when I have more time to focus on gettting the erlang track to v3… Perhaps after the pandemic…

Whatever you prefer!
